### PR TITLE
api: add `GET /api/v1/{owner}/{repo}/commits` endpoint

### PR DESCRIPTION
**Pull request by jaylevin**
This pull request targets issue #6573.

It provides a new API endpoint: `/api/v1/repos/{org}/{repo}/commits?pageSize=<int>` with a default page size of 30 commits (the same as the UI).

This implementation currently only focuses on the main/master branch of the repository, and does not provide the ability to return commit history for other branches.

- Note: Since the logic for converting a git.Commit to api.Commit had to be used in `GetAllCommits` and `GetSingleCommit`, I decided to pull the code out into a helper function, ` gitCommitToAPICommit(commit, context)`.

### CODE REVIEW

**Comment by jaylevin** [comment]
@unknwon anyway I can get a review/approval for this PR?

Best,
Jordan

File: internal/route/api/v1/repo/commits.go [hunk 1]
```diff
@@ -18,6 +19,52 @@ import (
 	"gogs.io/gogs/internal/gitutil"
 )
 
+// GetAllCommits returns a slice of commits starting from HEAD.
+func GetAllCommits(c *context.APIContext) {
+	if strings.Contains(c.Req.Header.Get("Accept"), api.MediaApplicationSHA) {
+		c.SetParams("*", c.Params(":sha"))
+		GetReferenceSHA(c)
+		return
+	}
```
Comment (unknwon) [inline internal/route/api/v1/repo/commits.go:28]: What is this part of the logic about?

File: internal/route/api/v1/repo/commits.go [hunk 2]
```diff
@@ -18,6 +19,52 @@ import (
...
+	// Get pagesize, set default if it is not specified.
+	pageSizeStr := c.Query("pageSize")
+	if pageSizeStr == "" {
+		pageSizeStr = "30"
+	}
+
+	pageSize, err := strconv.Atoi(pageSizeStr)
+	if err != nil {
+		c.Error(err, "get page size error")
+	}
```
Comment (unknwon) [inline internal/route/api/v1/repo/commits.go:39]: Pretty sure you can use `c.QueryInt` here.

File: internal/route/api/v1/repo/commits.go [hunk 3]
```diff
@@ -18,6 +19,52 @@ import (
...
+
+	gitRepo, err := git.Open(c.Repo.Repository.RepoPath())
+	if err != nil {
+		c.Error(err, "open repository")
+		return
+	}
+
+	// The response object returned as JSON
+	result := []*api.Commit{}
```
Comment (unknwon) [inline internal/route/api/v1/repo/commits.go:48]:
```suggestion
	result := make([]*api.Commit, 0, pageSize)
```

We can preallocate to the `pageSize`.

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the PR! Left few comments.

File: internal/route/api/v1/repo/commits.go (hunk 1 above)
Comment (jaylevin) [reply internal/route/api/v1/repo/commits.go:28]: I think I might have forgot to remove this after copying from GetSingleCommit.

**Comment by jaylevin** [comment]
> Thanks for the PR! Left few comments.

Thank you for the review! The comments should be addressed 👍

File: internal/route/api/v1/repo/commits.go [hunk 4]
```diff
@@ -36,23 +70,71 @@ func GetSingleCommit(c *context.APIContext) {
...
+	if err != nil {
+		c.NotFoundOrError(gitutil.NewError(err), "get reference commit ID")
+		return
+	}
+	c.PlainText(http.StatusOK, sha)
+}
+
+// gitCommitToApiCommit is a helper function to convert git commit object to API commit.
+func gitCommitToAPICommit(commit *git.Commit, c *context.APIContext) (*api.Commit, error) {
+
```
Comment (unknwon) [inline internal/route/api/v1/repo/commits.go:121]:
```suggestion
```

**Review by unknwon — APPROVED** [review]
Thanks for the followup!

**Comment by unknwon** [comment]
_The lint is failing for things that are not related to this PR._

**Pull request merged**